Feature/update readme (#36)
scalar-labs/cassy: docs/getting-started.md

# URI of a blob store to manage backup files 
scalar.cassy.server.storage_base_uri=s3://your-bucket

> # Enum to tell Cassy which cloud service to use (aws_s3, azure_blob)

Reviewer: ```suggestion
# Type of storage that Cassy stores backup files (aws_s3, azure_blob)
```

scalar.cassy.server.storage_type=aws_s3

# URL of JDBC for managing some metadata such as backup and restore histories
scalar.cassy.server.metadata_db_url=jdbc:sqlite:cassy.db

# URL of SRV record. It is used to know and pause Cassandra application nodes to take a cluster-wide snapshot. If you don't use the feature, it can be omitted or the value can be left blank.
scalar.cassy.server.srv_service_url=_app._tcp.your-service.com
```

> ### Cloud Storage Type Configuration

Reviewer: ```suggestion
### Storage type configuration
```
since ssh server is coming.

> To change the cloud service Cassy uses, the following properties of the configuration file will need to be modified:

Reviewer: ```suggestion
To change the storage that Cassy stores backup files, the following properties of the configuration file need to be modified:
```

> 1. `storage_base_uri`: the url to your S3 bucket or Azure blob.

Reviewer 2: In my understanding, `storage_base_uri` should be the container name if it is azure blob storage.
Please confirm

Author: Could you share your configuration file? I'm sorry, I should have left an example.

During my testing, I set `storage_base_uri = https://cassydev.blob.core.windows.net/indetail-cassy-test`

`indetail-cassy-test` is the name of the blob container.

Reviewer 2: My current working configuration
```
scalar.cassy.server.ssh_user=centos
scalar.cassy.server.ssh_private_key_path=/home/centos/.ssh/id_rsa
scalar.cassy.server.slave_command_path=/opt/cassy/build/install/cassy/bin
scalar.cassy.server.storage_base_uri=scalar-test
scalar.cassy.server.storage_type=AZURE_BLOB
scalar.cassy.server.metadata_db_url=jdbc:sqlite:/home/centos/cassy/cassy.db
scalar.cassy.server.srv_service_url=
```

Reviewer 2: I got a few exceptions when I used `scalar.cassy.server.storage_base_uri=https://cassydev.blob.core.windows.net/scalar-test`

```
Caused by: com.azure.storage.blob.models.BlobStorageException: Status code 400, "?<?xml version="1.0" encoding="utf-8"?><Error><Code>InvalidResourceName</Code><Message>The specifed resource name contains invalid characters.
RequestId:26918a13-601e-0015-7a13-030763000000
Time:2020-03-26T02:12:12.1356498Z</Message></Error>"
```

Author: @scalar-boney Sorry, I was wrong. We changed the code so it should only be the name of the container. `scalar.cassy.server.storage_base_uri=scalar-test` was correct. Does it work?

Reviewer 2: >  Does it work?
yes.

Sorry, 
Now I am facing another issue in the backup restore.
Now I have configured Cassandra and cassy in local machine but backup restore is failed.
I have attached  [test details herewith](https://docs.google.com/spreadsheets/d/10gBm2HdhkTibxUYwfhn9v6Qowo4kinZwfH5_h3_OzyQ/edit?usp=sharing)
Could you please check.

Reviewer 2: Sorry,
I have tried cassy with s3 storage but backup restore command failed.
Details attached [here](https://docs.google.com/spreadsheets/d/10gBm2HdhkTibxUYwfhn9v6Qowo4kinZwfH5_h3_OzyQ/edit?usp=sharing)

> 1. `storage_base_uri`: the url to your S3 bucket or Azure blob.

Reviewer: ```suggestion
1. `storage_base_uri`: the uri to your S3 bucket or Azure blob.
```
@yume-ind I'm just wondering, in Azure case, you said it should specify container name, but is there any reason why we can't specify the full URI ? `https://cassydev.blob.core.windows.net/scalar-test`
It just doesn't work ? 
Since it's currently named `storage_base_uri`, users will expect URI here, but if we need to specify a container name, it's a bit confusing for users, so we might need to rename it if we need to specify container name.

Reviewer 3: There is a way. 
If  `storage_base_uri=https://cassydev.blob.core.windows.net/scalar-test`, we can update the code to extract the container name from this URI like the following.
1. Verify the blob storage address `storage_base_uri` is the same as the one from the connection string environment variable, i.e. `https://cassydev.blob.core.windows.net`
2. Extract the container name from the `storage_base_uri` via string manipulation.

What do you think ?

Reviewer: @Torch3333 If we store the URI in the config, what is the point of having it in an environment variable ? 
Was there any particular reason of having `https://cassydev.blob.core.windows.net` in an env and `scalar-test` in the config separately ?

Reviewer 3: We choose to use the `connection string` to login (and which is passed as an env variable) because it seems the easiest way to set up authentication credentials for the blob. 
The `connection string` can be directly generated on the blob page. For example, for the `cassydev` blob, this will be here
https://portal.azure.com/#@scalar-labs.com/resource/subscriptions/b647bf0e-0566-4b28-ac8c-37b24f3b7061/resourceGroups/indetail/providers/Microsoft.Storage/storageAccounts/cassydev/keys
But the `connection string` only contains the credentials for a specific blob but does not point to a container of this blob, hence the user need to provide the `container` name separately that will be used to store the Cassandra snapshot. 
That's why, we used the `storage_base_uri` to pass the `container` name.
Asking the user to set these 2 parameters separately seemed to be the best compromise between usability and configurability.

Reviewer: OK, thank you for the detail explanation.
Let me make sure one thing. 
So, at the moment, we set `connection string` in an env manually, right ?

I think it would be a bit better for users to specify with URI format like this `storage_base_uri=https://cassydev.blob.core.windows.net/scalar-test`.

Reviewer 3: ```
So, at the moment, we set connection string in an env manually, right ?
```
Yes, the `connection_string` has to be set manually. 

```
I think it would be a bit better for users to specify with URI format like this storage_base_uri=https://cassydev.blob.core.windows.net/scalar-test.
```
Should I implement it as explained in this message ?
https://github.com/scalar-labs/cassy/pull/36#discussion_r399918842

Reviewer: Yes, please. it would be great if it is done in another PR.

Reviewer 3: Ok

2. `storage_type`: an Enum that tells Cassy which service to use. You can select `aws_s3` for Amazon S3, or `azure_blob` for Azure Blob.

> #### Azure-Specific Requirements

Reviewer: ```suggestion
#### Azure-specific requirements
```

To use Azure Blob, you will also need to set an environment variable on your system. Navigate to the Azure portal and [retrieve the connection string.](https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string#view-and-copy-a-connection-string)

Once you have copied the `connection_string`, use the following command to set your environment variable.
```

> export AZURE_STORAGE_CONNECTION_STRING=[your-connection-string]

Reviewer 2: It should be set in each Cassandra nodes also
Please confirm

Author: As far as I know, each slave node that will be using Azure will need to have this environment variable set. I will update the documentation to make this clearer.
